# Curso: Introducción a la Ciencia de Datos

**Ejercicio SF Salaries**
Utilizaremos el `SF Salaries Dataset` Kaggle. Simplemente siga y complete las tareas que se detallan a continuación en negrita. Las tareas serán cada vez más difíciles a medida que avance.

La idea principal es aplicar pandas para organizar información, manipularla, buscar información en el dataset


**Importe pandas como pd. e imprima la versión que está usando**

In [ ]:
import pandas as pd

In [2]:
pd.__version__

'2.3.3'

**Leer el archivo Salaries.csv en un dataframe llamado salarios**

Se encuentra en esta url y se puede leer directamente con pandas

https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv


In [ ]:
url = "https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv"

salarios = pd.read_csv(url)

salarios.head()


,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,NaN,San Francisco,NaN
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,NaN,San Francisco,NaN
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,NaN,San Francisco,NaN
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,NaN,San Francisco,NaN
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,NaN,San Francisco,NaN


**Use el método `.info()` para encontrar cuántas columnas de atributos hay y qué tipo de datos tiene cada una.**

In [ ]:
salarios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


- Según la información del dataframe su tamaño es de *memory usage: 14.7+ MB*

- Las columnas **Notes** y **Status** solamente tienen valores *NaN*, verifique qué dice **0 non-null** osea 0 valores son diferentes a *NaN*
entonces todos los valores son *NaN* en esas columnas

**Elimine las columnas *Notes* y *Status* Recuerde usar el método de forma permanente (inplace), y luego visualice cuántas filas y columnas tiene luego de eliminar las columnas**

In [ ]:
salarios.drop(columns=["Notes", "Status"], inplace=True)

In [6]:
salarios.shape

(148654, 11)

**¿Cuál es la nueva información del dataframe?**

In [ ]:
salarios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  object 
dtypes: float64(6), int64(2), object(3)
memory usage: 12.5+ MB


- Se puede verificar que al borrar las columnas que no tenían solo *NaN* se disminuye el tamaño del dataframe a: *memory usage: 12.5+ MB*

**Observe los primeros 5 registros del DataFrame**

In [ ]:
salarios.head()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,San Francisco


**Observe los últimos 5 registros del DataFrame**

In [ ]:
salarios.tail()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
148649,148650,Roy I Tillery,Custodian,0.0,0.0,0.00,0.0,0.00,0.00,2014,San Francisco
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148652,148653,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0.0,-618.13,-618.13,2014,San Francisco


**¿Cuántas filas y columnas tiene el dataframe?**

In [ ]:
salarios.shape

(148654, 11)

**Observe los 10 registros que están desde el índice 3000**

In [ ]:
salarios.iloc[3000:3010]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
3000,3001,WILLIAM BOYLE,FIREFIGHTER,105934.65,18530.98,16389.60,NaN,140855.23,140855.23,2011,San Francisco
3001,3002,ALAN LIN,ELECTRONIC MAINTENANCE TECHNICIAN,98965.80,41504.40,380.00,NaN,140850.20,140850.20,2011,San Francisco
3002,3003,WILLIAM CONLEY,SERGEANT III (POLICE DEPARTMENT),130457.78,3131.42,7256.39,NaN,140845.59,140845.59,2011,San Francisco
3003,3004,ROBERT MCDONALD,"LIEUTENANT, FIRE DEPARTMENT",123104.99,1833.56,15897.68,NaN,140836.23,140836.23,2011,San Francisco
3004,3005,MICHAEL COSTELLO,POLICE OFFICER III,112401.08,7957.74,20466.93,NaN,140825.75,140825.75,2011,San Francisco
3005,3006,EDNA HO,MANAGER V,140821.37,0.00,0.00,NaN,140821.37,140821.37,2011,San Francisco
3006,3007,RODRIGO CONEL,REGISTERED NURSE,126120.00,3981.78,10714.21,NaN,140815.99,140815.99,2011,San Francisco
3007,3008,VICTOR HURTADO,FIREFIGHTER,105934.67,13112.76,21763.50,NaN,140810.93,140810.93,2011,San Francisco
3008,3009,DONALD LANDRY,TRANSIT POWER LINE WORKER,92861.50,39789.32,8135.86,NaN,140786.68,140786.68,2011,San Francisco
3009,3010,FRANK LEE,"INSPECTOR III, (POLICE DEPARTMENT)",130481.56,725.40,9579.58,NaN,140786.54,140786.54,2011,San Francisco


La columna *Benefits* tiene valores `NaN`, ¿cuántos NaN tiene la columna *Benefits*? (TIP: Seleccione la Columna y vefique cuantos Na tiene `.isna()` luego sumelos `.sum()` )

In [ ]:
salarios["Benefits"].isna()

0          True
1          True
2          True
3          True
4          True
          ...  
148649    False
148650     True
148651     True
148652     True
148653    False
Name: Benefits, Length: 148654, dtype: bool

In [13]:
salarios["Benefits"].isna().sum()

np.int64(36163)

Al parecer los valores faltantes en la columna Benefits es por que no se le pagan beneficios entonces:

**Reemplace los valores *NaN* de la Columna *Benefits* por el valor de cero (Recuerde usar el método de forma inplace) y mostrar los primeros 10 elementos de la lista**

In [ ]:
salarios['Benefits'].fillna(0, inplace=True)

salarios['Benefits'].head(10)


C:\Users\im00795\AppData\Local\Temp\ipykernel_14748\1949094121.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  salarios['Benefits'].fillna(0, inplace=True)


0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
5    0.0
6    0.0
7    0.0
8    0.0
9    0.0
Name: Benefits, dtype: float64

**¿Cuál es la nueva información del dataframe?**

In [ ]:
salarios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  object 
dtypes: float64(6), int64(2), object(3)
memory usage: 12.5+ MB


**¿Cuántos valores únicos hay en las columnas categóricas?**

In [ ]:
salarios.select_dtypes(include='object').nunique()

EmployeeName    110811
JobTitle          2159
Agency               1
dtype: int64

- la columna **Agency** solamente tiene un valor único, entonces no aporta información, por lo tanto se puede eliminar

**Elimine la columna *Agency* del dataframe**

In [17]:
salarios.drop('Agency', axis=1, inplace=True)

**¿Cuál es el promedio de BasePay?**

In [18]:
salarios['BasePay'].mean()

np.float64(66325.44884048769)

**¿Cuál es la cantidad más alta de OvertimePay en el dataset?**

In [19]:
salarios['OvertimePay'].max()

np.float64(245131.88)

**Realice una descripción estadística de las columnas numéricas**

In [20]:
salarios.describe()

,Id,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year
count,148654.000000,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000,148654.000000,148654.000000
mean,74327.500000,66325.448840,5066.059886,3648.767297,18924.232839,74768.321972,93692.554811,2012.522643
std,42912.857795,42764.635495,11454.380559,8056.601866,17165.279495,50517.005274,62793.533483,1.117538
min,1.000000,-166.010000,-0.010000,-7058.590000,-33.890000,-618.130000,-618.130000,2011.000000
25%,37164.250000,33588.200000,0.000000,0.000000,0.000000,36168.995000,44065.650000,2012.000000
50%,74327.500000,65007.450000,0.000000,811.270000,23214.020000,71426.610000,92404.090000,2013.000000
75%,111490.750000,94691.050000,4658.175000,4236.065000,33468.980000,105839.135000,132876.450000,2014.000000
max,148654.000000,319275.010000,245131.880000,400184.250000,96570.660000,567595.430000,567595.430000,2014.000000


**En este dataset las variables `Id` y `Year` representan variables categóricas, convertir estas columnas a su formato correcto**

In [21]:
salarios['Id'] = salarios['Id'].astype('category')

salarios['Year'] = salarios['Year'].astype('category')

**La variable `Year` es categórica ordinal, primero detecte cuáles son los valores únicos que tiene la columna**

In [22]:
salarios['Year'].unique()

[2011, 2012, 2013, 2014]
Categories (4, int64): [2011, 2012, 2013, 2014]

**A partir de la información anterior convierta la columna `Year` en variable categórica ordinal y visualice la columna**

In [23]:
salarios['Year'] = pd.Categorical(
    salarios['Year'],
    categories=[2011, 2012, 2013, 2014],
    ordered=True)

In [24]:
salarios['Year']

0         2011
1         2011
2         2011
3         2011
4         2011
          ... 
148649    2014
148650    2014
148651    2014
148652    2014
148653    2014
Name: Year, Length: 148654, dtype: category
Categories (4, int64): [2011 < 2012 < 2013 < 2014]

**Realice nuevamente una descripción estadística de las columnas numéricas**

In [25]:
salarios.describe()

,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits
count,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000,148654.000000
mean,66325.448840,5066.059886,3648.767297,18924.232839,74768.321972,93692.554811
std,42764.635495,11454.380559,8056.601866,17165.279495,50517.005274,62793.533483
min,-166.010000,-0.010000,-7058.590000,-33.890000,-618.130000,-618.130000
25%,33588.200000,0.000000,0.000000,0.000000,36168.995000,44065.650000
50%,65007.450000,0.000000,811.270000,23214.020000,71426.610000,92404.090000
75%,94691.050000,4658.175000,4236.065000,33468.980000,105839.135000,132876.450000
max,319275.010000,245131.880000,400184.250000,96570.660000,567595.430000,567595.430000


**¿Cuál es el título de trabajo de JOSEPH DRISCOLL? Nota: Utilice mayúsculas, de lo contrario, puede obtener una respuesta que no coincida**

In [26]:
salarios[salarios['EmployeeName'] == 'JOSEPH DRISCOLL']['JobTitle']

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

**¿Cuánto gana JOSEPH DRISCOLL (incluyendo beneficios)?**

In [27]:
salarios[salarios['EmployeeName'] == 'JOSEPH DRISCOLL']['TotalPayBenefits']

24    270324.91
Name: TotalPayBenefits, dtype: float64

**¿Cuál es el nombre de la persona con mejor pago (incluyendo beneficios)?**

In [28]:
salarios.loc[salarios['TotalPayBenefits'].idxmax()][['EmployeeName','TotalPayBenefits']]

EmployeeName        NATHANIEL FORD
TotalPayBenefits         567595.43
Name: 0, dtype: object

**¿Cuál es el nombre de la persona peor paga (incluyendo beneficios)? ¿Notan algo extraño sobre cuánto le pagan?**

In [29]:
salarios.loc[salarios['TotalPayBenefits'].idxmin()][['EmployeeName','TotalPayBenefits']]

EmployeeName        Joe Lopez
TotalPayBenefits      -618.13
Name: 148653, dtype: object

Es extraño porque es un valor negativo, es decir, saldo a favor de la empresa. Como si al empleado se le realizara un ajuste por una deuda. O puede ser también un error en nómina.

**¿Cuántos registros tienen un valor de TotalPayBeneficts menor que cero?**

In [30]:
(salarios['TotalPayBenefits'] < 0).sum()

np.int64(4)

**¿Cuál es el promedio (mean) BasePay de todos los empleados por año? los años van entre (2011-2014)?** (Tip: Usar groupby)

In [31]:
salarios.groupby('Year')['BasePay'].mean()

C:\Users\im00795\AppData\Local\Temp\ipykernel_14748\1090324270.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  salarios.groupby('Year')['BasePay'].mean()


Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

**¿Cuántos titulos de trabajos (JobTitle) únicos existen?**

In [32]:
salarios['JobTitle'].nunique()

2159

**¿Cuáles son los 5 trabajos más comunes?**

In [33]:
salarios['JobTitle'].value_counts().head(5)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

**¿Cuántos puestos de trabajo únicos existieron en 2013? (Ejemplo, los títulos de trabajo con solo una ocurrencia en 2013?)**

In [34]:
salarios_2013 = salarios[salarios['Year'] == 2013]

(salarios_2013['JobTitle'].value_counts() == 1).sum()

np.int64(202)

**¿Cuántas personas tienen la palabra Chief en el título de su trabajo?** Apoyese en los métodos para manipular strings y en el método `.apply()`

In [35]:
salarios['JobTitle'].apply(lambda x: 'chief' in x.lower()).sum()

np.int64(627)

Este código es válido, se recomienda hacer el código mas simple y agregar typing y documentación.

In [36]:
def contains_chief(title: str) -> bool:
    return 'chief' in title.lower()

salarios['JobTitle'].apply(contains_chief).sum()

np.int64(627)

**Bono: ¿Existe una correlación entre la longitud del nombre del Título del trabajo y el Salario?**

Crea una nueva columna donde se guarde la longitud del titulo del trabajo. Tip: usar el método `.apply()`

In [38]:
salarios['JobTitle_len'] = salarios['JobTitle'].apply(len)

salarios['JobTitle_len'].head()

0    46
1    31
2    31
3    36
4    44
Name: JobTitle_len, dtype: int64

Utilice el método `.corr()` de Pandas para calcular la correlación

In [39]:
salarios['JobTitle_len'].corr(salarios['TotalPayBenefits'])

np.float64(-0.03687844593260669)

Esta correlación, al ser negativa y cercana a cero, indica que la relación entre la longitud del nombre del Título y el Salario es inversa (tener un título largo no implica ganar más dinero, y en sentido contrario) y muy débil, casi inexistente. En conclusión, el salario depende de otros factores.  